### ISSUE TITLE
Potential performance improvement: Turn partial publish into full publish if outcome is the same

### ISSUE THREAD

**Issue by bwaidelich** [neos/neos-development-collection#5303]
*TL;DR:* A partial publish is much slower and more complex than a full publish of a workspace – we can detect if a partial publish already contains all pending events and turn it into a full publish in this case.

For Neos the default publishing mode is to only publish events of the currently selected document (IMHO a full publish should be the default but that discussion is out of scope!).
A partial publish is much more complex than a full publish because we first have to split affected and remaining events and publish those to different newly forked streams...

During the splitting of those events we could handle the `PublishIndividualNodesFromWorkspace` like the corresponding `PublishWorkspace` command if `$remainingCommands` is empty.

The same should be true for `DiscardIndividualNodesFromWorkspace` vs `DiscardWorkspace`

Related: #4388

**Comment by mhsdesign** [comment]
jup @kitsunet and me had the same thoughts :) it will be implemented via https://github.com/neos/neos-development-collection/pull/5301

**Comment by mhsdesign** [comment]
Funnily after thinking about this https://github.com/neos/neos-development-collection/pull/5301 doesnt really solve this case really. As a full publish will with that pr also do the simulation and apply the commands. We only said "do a full publish by emitting `WorkspaceWasPublished` this is simpler for the projections to handle" which is not true either. The projections handle the events the same. So just for completion: We always have deserialise all the commands that need to be reapplied and splitting them by ids doesnt cost anything. The costly simulation will be carried out for both partial and full publish.

And just for completeness with https://github.com/neos/neos-development-collection/pull/5385 will the `WorkspaceWasPartially` published event actually removed as we only need a `WorkspaceWasPublished` event at all.

**Comment by bwaidelich** [comment]
> We only said "do a full publish by emitting WorkspaceWasPublished this is simpler for the projections to handle"

I don't know who said that, but it wasn't me. My point was: If is much easier to reason about the events if we can eliminate the last case of:
- all changes were published
- some changes were published
- all changes were published but through a different button

> with https://github.com/neos/neos-development-collection/pull/5385 will the WorkspaceWasPartially published event actually removed

That would be the great ofc, but I really don't understand how #5385 is supposed to work. And IMO the difference between a partial and a full publish has a different meaning and should not be distinguished just by some metadata

**Comment by mhsdesign** [comment]
Just to close this up: The partial distinction is now still made via flag: https://github.com/neos/neos-development-collection/pull/5385#issuecomment-2532238085, but there is no performance gain whatsoever - as we changed the rules of publish and made PublishAll equally slow as PublishPartially :D
